pyquil/quil.py

```diff
@@ -719,15 +719,59 @@ def instantiate_labels(instructions):
     return result
 
 
+def merge_with_pauli_noise(prog_list, probabilities, qubits):
+    """
+    Insert pauli noise channels between each item in the list of programs.
+    This noise channel is implemented as a single noisy gate acting on the provided qubits.
+
+    :param prog_list: an iterable such as a program or a list of programs.
+    If a program is provided, a single noise gate will be applied after each gate in the program.
+    If a list of programs is provided, the noise gate will be applied after each program.
+    :param list|floats probabilities: The 4^num_qubits list of probabilities specifying the desired pauli channel.
+    There should be either 4 or 16 probabilities specified in the order I,X,Y,Z or II, IX, IY, IZ, XI, XX, XY, etc
+    :param list qubits: a list of the qubits that the noisy gate should act on.
+    :return: A single program with noisy gates inserted between each element of the program list.
+    :rtype: Program
+    """
+    p = Program()
+    p.defgate("pauli_noise", np.eye(2 ** len(qubits)))
+    p.define_noisy_gate("pauli_noise", qubits, pauli_kraus_map(probabilities))
+    for elem in prog_list:
+        p.inst(Program(elem)).inst(("pauli_noise", *qubits))
+    return p
+
+
 def merge_programs(prog_list):
     """
-    Merges a list of pyQuil programs into a single one by appending them in sequence
+    Merges a list of pyQuil programs into a single one by appending them in sequence.
+    If multiple programs in the list contain the same gate and/or noisy gate definition
+    with identical name, this definition will only be applied once. If different definitions
+    with the same name appear multiple times in the program list, each will be applied once
```

Comment:
Why do you apply things with the same name? Won't the last one be the one that is used anyway?

Reply:
I was thinking of the edits to merge_programs as being an alternative to merge_with_pauli_noise. Changing merge_programs makes the most sense for the RB code, but I'm not sure if this is the way we should change pyquil. The usage would be something like
```python
program_list = generate_rb_sequence() 
for p in program_list:
  p.defgate("pauli_noise", np.eye(2 ** len(qubits)))
  p.define_noisy_gate("pauli_noise", qubits, pauli_kraus_map(probabilities))
  p.inst("pauli_noise")
final = merge(program_list)
```

If you try this with the old merge_programs you get a warning because there is a defgate in the final merged_program for each program in the list. I didn't find a way to avoid redundant Kraus operators. 

I found that the last defgate is the one that is used, but the first define_noisy_gate is used. I wanted to avoid applying truly redundant defgates, but if there were conflicting definitions apply them in the order they last appear so that the behavior is the same as the old merge_programs (the last gate definition is taken and a warning is thrown).

Reply:
That all sounds good, let me make sure I understand a couple things:

First, you're saying you can implement `merge_with_pauli_noise` by using `merge_programs`? That sounds good to me. Especially since you went through the work of getting `merge_programs` to be correct. :)

Second, it sounds like you're saying reason you chose to apply things the way you did is to ensure backwards compatibility. That sounds good to me as well. I guess my only other point is that if two definitions have the same name, you don't need to store all of them, do you? Can't you just go through the list once in forward order, holding onto the most recently seen definition for a given name?

Reply:
I could implement merge_with_pauli_noise using merge_programs; this would result in a lot of redundant Kraus Pragmas, but if it's worth the code simplicity then I'm fine with that (do redundant pragmas impact performance besides increasing the size of the program object)? 

I'm looking for backward compatibility with the warnings as well. If multiple different definitions with the same name occur, I want to include those in the final program so that the user knows they've included contradictory definitions and that only the last definition is being applied.

Reply:
That sounds fine to me. :+1: